SG-8048: App with devtools

manneohrstrom opened the pull request:
Adds new developer options as context menu actions:

![image](https://user-images.githubusercontent.com/337710/43876623-ebe778d0-9bd0-11e8-8add-2832924c55e1.png)

These appear dynamically depending on your config.

- if you are on a centralized config, nothing appears
- if you are on a distributed non-dev (e.g. zipped) config, you get the option to create a config sandbox.
- once you go into the dev sandbox (e.g. configs with dev descriptor), you get two additional options that let you show the config in the file system as well as run `tank updates` on it.


The create dev sandbox dialog looks like this:
![image](https://user-images.githubusercontent.com/337710/43876739-5ca051c8-9bd1-11e8-87f7-eb5a081375c3.png)

The check for updates looks like this:
![image](https://user-images.githubusercontent.com/337710/43876719-466d2d4a-9bd1-11e8-86a3-6617fe627850.png)

jfboismenu commented on app.py:
```diff
@@ -0,0 +1,124 @@
+        """
+        Called as the application is being initialized
+        """
+        pipeline_config_id = self.sgtk.pipeline_configuration.get_shotgun_id()
```
You are accessing a private api here. We should find a way to make this public.

manneohrstrom replied:
yes, let's discuss this. Maybe we need to make `pipeline_configuration` official?

jfboismenu commented on app.py:
```diff
@@ -0,0 +1,124 @@
+            ["windows_path", "linux_path", "mac_path"]
+        )
+
+        if sg_data and (sg_data["windows_path"] or sg_data["mac_path"] or sg_data["linux_path"]):
```
Could you make sure that is indeed the case? I'm pretty sure we also support the _path fields when using a plugin id. Not sure what it means tough. Anyhow, my point is that we have a method on the Toolkit instance that allows us to know if the configuration is centralized or not, so the logic is encapsulated there, but also because there's some confusion (I might be the one being confused actually!) over what constitutes as a centralized config.

jfboismenu replied:
I just looked and there seems to be an [oversight](https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/bootstrap/resolver.py#L375-L398).

We're assuming that if there is a path that a configuration is always centralized. I don't have a problem with that if we really want that to be the case. But then at the very least we should detect the error if both the path AND the plugin id set. We should log and skip it basically.

jfboismenu commented on app.py:
```diff
@@ -0,0 +1,124 @@
+
+        # certain menu options are only available if you
+        # are running out of a dev descriptor config:
+        if self.sgtk.configuration_descriptor.is_dev():
```
What is the difference between a dev and a path descriptor? I'm personally always using a path descriptor. Is that distinction really useful? Should we fold the concepts together?

manneohrstrom replied:
We used to have a dev descriptor only - intended for development only! then people started to use it to describe paths on disk to stuff for use in non-dev situations, so we added a path descriptor in 0.18. I think it makes sense to have both, it's useful to be able to say "when in dev mode, do this" - like this in this example. You should switch to using dev descriptors for your development work :)

jfboismenu commented on app.py:
```diff
@@ -0,0 +1,124 @@
+        """
+        disk_location = self.sgtk.configuration_descriptor.get_path()
+
+        if sys.platform.startswith("linux"):
```
This is a pattern that I've seen reproduced quite often and we should stop propagating it.

Could you please introduce a method in core that does the following:

```
def open_file(path):
    try:
        from sgtk.platform.qt import QDesktopServices
    except ImportError:
        # Do that logic you just typed
    else:
        QDesktopServices.openUrl(path)
```

Qt already has this logic built-in

manneohrstrom replied:
yes, it's duplicated in a bunch of places. I'll sort it out as part of my changes to core in the other PR.

manneohrstrom replied:
done. in 18.156

jfboismenu commented on info.yml:
```diff
@@ -0,0 +1,33 @@
+
+# Required minimum versions for this item to run
+requires_shotgun_version:
+requires_core_version: "v0.18.122"
```
We'll want to bump this version number after we introduce the changes I've suggested to core.

manneohrstrom replied:
good point!

jfboismenu commented on python/__init__.py:
```diff
@@ -0,0 +1,12 @@
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+from . import create_sandbox
+from . import check_updates
```
W00f (I'm going to pretend I'm Hound CI). Missing end of file CR.

jfboismenu commented on python/check_updates/__init__.py:
```diff
@@ -0,0 +1,27 @@
+
+    # start ui
+    app.engine.show_dialog("Check for updates", app, AppDialog)
+
```
Woof. Emtpy lines at the end of the file.

jfboismenu commented on python/check_updates/dialog.py:
```diff
@@ -0,0 +1,184 @@
+
+        :return: a string for the user input.
+        """
+        if "[Yna?]" in self._previous_stdout_stderr_line:
```
I don't get this. Are you saying that we are capturing stdout and popping message boxes to ask for yes/no/all?

It feels like we should instead have a callback that you can pass to the interactive command that receives in the question and the list of choices you can make. Then we built a UI and return the result. The default implementation would simply use stdin. Parsing stdout seems like we're going to break this at some point.

manneohrstrom replied:
Let's discuss the pragmatic way forward here - these prompts are generated by a [standard set of methods](https://github.com/shotgunsoftware/tk-core/blob/master/python/tank/commands/console_utils.py#L45) in core so it's not random. But yes, we are parsing stdout.

jfboismenu replied:
Which are not unit tested, so we can break this. ;)

I simply feel that the number of lines of code to use callbacks in core is going to much lower than the number of lines in your app to capture and write text. Callbacks are going to be easier to maintain and easier to understand and not prone to breakage if we change what text is output.

jfboismenu commented on python/check_updates/redirect.py:
```diff
@@ -0,0 +1,156 @@
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sys
```
All this complexity would go away if we simply passed in a callback to the command tank command.

jfboismenu commented on python/check_updates/yes_no_dialog.py:
```diff
@@ -0,0 +1,77 @@
+
+import sgtk
+
+from sgtk.platform.qt import QtCore, QtGui
```
Without a screenshot, it's hard to tell what is the difference here between this and a QMessageBox, especially since QMesssageBox supports a Yes/No/Yes to All. If even supports buttons with custom labels if you really want to use `Always` as a caption.

Could we use a simple QMessageBox?

jfboismenu commented on python/create_sandbox/__init__.py:
```diff
@@ -0,0 +1,27 @@
+
+    # start ui
+    app.engine.show_dialog("Create Config Sandbox", app, AppDialog)
+
```
Woof. Empty lines.

jfboismenu commented on python/create_sandbox/dialog.py:
```diff
@@ -0,0 +1,176 @@
+                logger.debug("Copying config files from %r to %s" % (config_descriptor, path))
+                config_descriptor.copy(path)
+                # delete system files
+                sgtk.util.filesystem.safe_delete_folder(
```
Should this be part of the copy?

jfboismenu commented on python/create_sandbox/dialog.py:
```diff
@@ -0,0 +1,176 @@
+            # get the current pipeline config id
+            sg_data_current_pc = self._bundle.shotgun.find_one(
+                "PipelineConfiguration",
+                [["id", "is", self._bundle.sgtk.pipeline_configuration.get_shotgun_id()]],
```
Access to private API.

jfboismenu replied:
Maybe we need `self._bundle.sgtk.get_pipeline_configuration_entity()` which returns a dict with keys `name`, `plugin_id`, `descriptor`, `*_path` and does the massaging to we don't have to deal with sg_plugin_ids vs plugin_ids and sg_description vs descriptor.

Emphasis on the `_entity` bit in case in the future we decide to have an actual pipeline configuration object in the public API.

jfboismenu commented on python/create_sandbox/dialog.py:
```diff
@@ -0,0 +1,176 @@
+
+            QtCore.QCoreApplication.processEvents()
+
+            descriptor_uri = "sgtk:descriptor:dev?path=%s" % urllib.quote(path)
```
Shouldn't we set the platform's field instead of the general path field? Setting the platform field would give a
"This descriptor is not configured for this current OS." like error in the resolver instead of "Path does not exists."

manneohrstrom replied:
cool idea!

jfboismenu commented on python/create_sandbox/dialog.py:
```diff
@@ -0,0 +1,176 @@
+
+            # ok we are good to go!
+            logger.debug("Creating new pipeline config in Shotgun...")
+            sg_data = self._bundle.shotgun.create(
```
We should make it clear at this point that the sandbox has been created, but that we haven't switched to it and that the user needs to restart the DCC or exit and reenter their project in Desktop.

manneohrstrom replied:
it currently says:

![image](https://user-images.githubusercontent.com/337710/46491401-8f08ab80-c802-11e8-9482-c7bdf7390f2e.png)

jfboismenu commented on resources/create_sandbox/build_resources.sh:
```diff
@@ -0,0 +1,47 @@
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+# The path to where the PySide binaries are installed
+PYTHON_BASE="/Applications/Shotgun.app/Contents/Resources/Python/bin"
```
I would so love to refactor this piece of logic into a tk-build python module or something....

manneohrstrom replied:
oh yes!

manneohrstrom replied:
FYI i have created a separate ticket for this work.

jfboismenu commented on python/create_sandbox/dialog.py:
```diff
@@ -0,0 +1,176 @@
+        self.ui.setupUi(self)
+
+        # untick copy checkbox
+        self.ui.copy_config.setChecked(False)
```
I think that checkbox should probably on by default. If you are creating a sandbox, it's usually to tweak things.

manneohrstrom replied:
see logic below. if the folder specified contains any files, the checkbox gets unticked, otherwise it is ticked. The rationale is that if there is stuff already there in the location we are setting up, we default to not copy. If the folder is empty, we default to copy.

manneohrstrom commented:
Ready for another round of review. Based on your feedback, refactored the interaction with the tank command (FWIW, turned this into a much bigger chunk of work) and added explicit methods to core rather than accessing `pipeline_configuration`.

manneohrstrom commented on app.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        Called as the application is being initialized
+        """
+        if self.sgtk.configuration_mode != self.sgtk.DISTRIBUTED:
```
logic for determining the config mode is now in core.

manneohrstrom commented on app.py:
```diff
@@ -0,0 +1,125 @@
+        """
+        Callback to launch the validate tank command
+        """
+        command_runner = self.import_module("command_runner")
```
turned updates ui into a generic tank command runner.

manneohrstrom commented on info.yml:
```diff
@@ -0,0 +1,33 @@
+
+# Required minimum versions for this item to run
+requires_shotgun_version:
+requires_core_version: "v0.18.156"
```
i will bump this later on.

manneohrstrom commented on python/command_runner/dialog.py:
```diff
@@ -0,0 +1,81 @@
+        self._logger.addHandler(self._handler)
+
+        # set up interaction interface
+        self._interaction_interface = QtInteraction(self)
```
new interface for interacting with tank commands

manneohrstrom commented on python/command_runner/dialog.py:
```diff
@@ -0,0 +1,81 @@
+        Cancel execution
+        """
+        if self._command_obj:
+            self._command_obj.terminate()
```
new method for requesting a termination of tank commands.

manneohrstrom commented on python/create_sandbox/dialog.py:
```diff
@@ -0,0 +1,177 @@
+            # get the current pipeline config id
+            sg_data_current_pc = self._bundle.shotgun.find_one(
+                "PipelineConfiguration",
+                [["id", "is", self._bundle.sgtk.configuration_id]],
```
new property in core.
